src/encoded/ingestion/structured_data.py

```diff
@@ -0,0 +1,606 @@
+            array_name, array_index = _StructuredRowData._get_array_info(column_name_component)
+            name = array_name if array_name else column_name_component
+            if len(column_name_components) > 1:
+                if not isinstance(row[name], dict) and not isinstance(row[name], list):
+                    row[name] = {}
+                setv(row[name], column_name_components[1:], parent_array_index=array_index)
+                return
+
+            nonlocal column_name, value, schema, loc
+            if schema:
+                value = schema.map_value(value, column_name, loc)
+            if array_name is not None and isinstance(value, str):
+                value = split_array_string(value)
+            if array_name and array_index >= 0:
+                if isinstance(row[name], str):  # An array afterall e.g.: abc,abc#2
+                    row[name] = split_array_string(row[name])
+                if len(row[name]) < array_index + 1:
+                    row[name].extend([None] * (array_index + 1 - len(row[name])))
+                row[name] = row[name][:array_index] + value + row[name][array_index + 1:]
+            else:
+                row[name] = value
+
+        setv(row, split_dotted_string(column_name))
+
+    @staticmethod
+    def _parse_into_row_template(column_names: List[str]) -> dict:
+
+        def parse_components(column_name_components: List[str]) -> dict:
+            value = parse_components(column_name_components[1:]) if len(column_name_components) > 1 else None
+            column_name_component = column_name_components[0]
+            array_name, array_index = _StructuredRowData._get_array_info(column_name_component)
+            if array_name:
+                array_length = array_index + 1 if array_index >= 0 else (0 if value is None else 1)
+                # Doing it the obvious way, like in the comment right below here, we get
+                # identical (shared) values; which we do not want; so do a real/deep copy.
+                # return {array_name: [value] * array_length}
+                return {array_name: [copy.deepcopy(value) for _ in range(array_length)]}
+            return {column_name_component: value}
+
+        structured_row_template = {}
+        for column_name in column_names or []:
+            if (column_name_components := split_dotted_string(column_name)):
+                merge_objects(structured_row_template, parse_components(column_name_components), True)
+        return structured_row_template
+
+    @staticmethod
+    def _get_array_info(name: str) -> Tuple[Optional[str], Optional[int]]:
+        if (array_indicator_position := name.rfind(ARRAY_NAME_SUFFIX_CHAR)) > 0:
+            array_index = name[array_indicator_position + 1:] if array_indicator_position < len(name) - 1 else -1
+            if (array_index := to_integer(array_index)) is not None:
+                return name[0:array_indicator_position], array_index
+        return None, None
+
+
+class Schema:
```

Comment:
Is this class at all necessary? What does this do that `load_schema` doesn't?

Reply:
The main point of this class is to create/contain the stripped-down "flattened" version of the schema for quick lookup of types to convert values in the data appropriate to the schema (via map_value method); and also to validate data using the loaded schema (via validate method). But yes the part that gets/loads the schema uses the ff_utils.get_schema method, which ultimately does end up using/calling (snovault) load_schema.